### ISSUE TITLE
Frequent CoreExceptions

### ISSUE THREAD

**Issue by bkilshaw** [auth0/laravel-auth0#216]
### Describe the problem

We have a relatively high traffic application and are using Auth0 for authentication using our customers Ping Federate server. We are regularly receiving CoreExceptions "Invalid State" when users login. I'm not entirely sure how they eventually get in, but I'm guessing they hit the login page and try again and it works. Auth0 logs are showing a successful login.

### What was the expected behavior?

Users should be able to login with no problem.

### Reproduction
Unable to reproduce as it's an intermittent issue and I can't ask a bunch of customers to login on a test site until it happens

### Environment
- **Version of this library used:** auth0/login: 6.4.0
- **Which framework are you using, if applicable:** Laravel 8.37.0

This is also happening on an older environment running:

auth0/login: 5.4.0
laravel: 7.28.4

My Auth0CallbackController and CustomUserRepository can be found here: https://gist.github.com/bkilshaw/b1c51841469cc2b2268fbe23d22e3254

**Comment by evansims** [comment]
Hey, @bkilshaw 👋 Sorry to hear you're hitting problems with this.

An invalid state core exception is [raised by the underlying PHP SDK](https://github.com/auth0/auth0-PHP/blob/master/src/Auth0.php#L550) when [the state set by the SDK](https://github.com/auth0/auth0-PHP/blob/master/src/Auth0.php#L424-L430) during [the login request](https://github.com/auth0/laravel-auth0/blob/master/src/Auth0/Login/Auth0Service.php#L99) differs from what it sees when the authentication flow finishes and is verified. This state is stored by the SDK on the requesting device as a cookie.

My first suggestion would be for us to look for some sort of a cookie-related storage issue happening. Do you have access to error logs with device metrics to see if there are specific browser types that are hitting this? Perhaps there's a common thread there we can start from.

**Comment by bkilshaw** [comment]
> Hey, @bkilshaw 👋 Sorry to hear you're hitting problems with this.
> 
> 
> 
> An invalid state core exception is [raised by the underlying PHP SDK](https://github.com/auth0/auth0-PHP/blob/master/src/Auth0.php#L550) when [the state set by the SDK](https://github.com/auth0/auth0-PHP/blob/master/src/Auth0.php#L424-L430) during [the login request](https://github.com/auth0/laravel-auth0/blob/master/src/Auth0/Login/Auth0Service.php#L99) differs from what [it sees](https://github.com/auth0/auth0-PHP/blob/master/src/Auth0.php#L549) when the authentication flow finishes and the state is [compared](https://github.com/auth0/auth0-PHP/blob/978ee208377c6f0e811a17ebd41bf01ef306e23d/src/Helpers/TransientStoreHandler.php#L94). This state is stored by the SDK on the requesting device with [a cookie](https://github.com/auth0/auth0-PHP/blob/978ee208377c6f0e811a17ebd41bf01ef306e23d/src/Store/CookieStore.php).
> 
> 
> 
> My first suggestion would be for us to look for some sort of a cookie-related storage issue happening. Do you have access to error logs with device metrics to see if there are specific browser types that are hitting this? Perhaps there's a common thread there we can start from.

I have full access to the server logs but nothing from the device itself. Is there anything I can log when the error occurs to help identify?

**Comment by evansims** [comment]
Would you be able to capture user agent strings to see if there's a commonality between the type of browser being used for users hitting that?

**Comment by bkilshaw** [comment]
> Would you be able to capture user agent strings to see if there's a commonality between the type of browser being used for users hitting that issue?

Alright I've started logging that and will update the issue once we have a few examples to compare.

**Comment by bkilshaw** [comment]
Nothing exhaustive yet, but so far I've logged the following:

` Apr 30 11:18:02 ["useragent","Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
Apr 30 11:18:03 ["useragent","Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
Apr 30 11:19:13 ["useragent","Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
Apr 30 11:29:42 ["useragent","Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
Apr 30 13:48:13 ["useragent","Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
Apr 30 13:49:08 ["useragent","Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
Apr 30 15:19:38 ["useragent","Mozilla/5.0 (iPad; CPU OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
Apr 30 16:25:51 ["useragent","Mozilla/5.0 (iPad; CPU OS 12_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
May 01 08:34:22 ["useragent","Mozilla/5.0 (iPad; CPU OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] 
May 01 11:50:35 ["useragent","Mozilla/5.0 (iPad; CPU OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1"] `

**Comment by evansims** [comment]
Thanks, @bkilshaw that's looking promising, please keep an eye on it, and let's see if that trend of the problem being mobile safari holds. If you have access to any iOS devices on your team that might be able to dogfood that too, particularly in different iOS versions, that might be useful to see if they're also affected.

**Comment by evansims** [comment]
Hey again, @bkilshaw 👋  Just wanted to check in and see if anything new had made its way into that log

**Comment by bkilshaw** [comment]
Hey @evansims 

Looks like these logs flush every 48 hours so I don't have a good history, but here are the user agents I've collected so far:

| Occurances  | Useragent |
| ------------- | ------------- |
| 8  | Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1  |
| 9  | Mozilla/5.0 (iPad; CPU OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1  |
| 1  | Mozilla/5.0 (iPad; CPU OS 12_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1  |

**Comment by evansims** [comment]
Thanks, @bkilshaw! So far this is feeling like an isolated situation, and I wonder if this is all from the same user's devices that they've perhaps done something like block cookies in their Safari privacy settings, or perhaps they're using an overzealous content blocker or filtering VPN. Let's keep an eye on it a bit longer and see if anything new finds its way into your logs

**Comment by evansims** [comment]
Hey @bkilshaw just wanted to check in again and see if you noticed any new user agent patterns in the logs

**Comment by bkilshaw** [comment]
@evansims From the last 48 hours:

| Occurrences  | Useragent |
| ------------- | ------------- |
|6 | Mozilla/5.0 (iPad; CPU OS 12_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1 |
|4 | Mozilla/5.0 (iPad; CPU OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1 |
|4 | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Safari/605.1.15 |
|3 | Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148 Safari/604.1 |
|1 | Mozilla/5.0 (iPhone; CPU iPhone OS 14_4_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1 |

**Comment by bkilshaw** [comment]
I know it's happened to multiple different users. Early on I got a few reports from users but didn't collect device information. I haven't received reports from users since our error page simply says to try and login again. Based on the date ranges it seems like people will receive the same error a couple times in a row on occasion, then I can't tell if they're getting in successfully or giving up, but I would assume the former otherwise they'd complain.

The app is used in a retail environment in a few dozen stores across the country. All store should have the same network setup as they're all owned by a single company. The app is generally accessed by all types of devices and browsers; PC's, personal phones, and the iPhones the store provides. Unfortunately I can't think of any way to figure out exactly what devices might be throwing the error other than the user-agent, but if you have any recommendations on stuff I can collect server side I'm all ears!

**Comment by bkilshaw** [comment]
@evansims Anything else I can look into now that we've identified it's happening on various devices?

**Comment by bkilshaw** [comment]
@evansims Also just for an update, last 48 hours we have logged a few errors with the following user agents:





<meta name="ProgId" content="Excel.Sheet">
<meta name="Generator" content="Microsoft Excel 15">
<link id="Main-File" rel="Main-File" href="file:////Users/bradkilshaw/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel="File-List" href="file:////Users/bradkilshaw/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>





Mozilla/5.0   (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko)   Version/14.0.3 Safari/605.1.15
--
Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1   like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1   Mobile/15E148 Safari/604.1
Mozilla/5.0 (iPhone; CPU iPhone OS 14_4_2   like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3   Mobile/15E148 Safari/604.1
Mozilla/5.0 (iPad; CPU OS 12_3_1 like Mac   OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.1.1 Mobile/15E148   Safari/604.1

**Comment by evansims** [comment]
Hey @bkilshaw 👋  So, I took some time to try to recreate this issue on my end over the holiday weekend using the devices I have access to running a few different MacOS majors and iOS Simulator, but unfortunately I'm just not able to reproduce this one. My guess would be it has something to do with a combination of Safari's cookie blocking behavior and the app environment, but I'm honestly at a loss to suggest what that could be. At this point I would suggest moving through [our support channel](https://support.auth0.com/) where one of our support engineers might be able to diagnose this one more effectively for you than I can

**Comment by bkilshaw** [comment]
@evansims I -think- I might have figured out what might causing the CoreExceptions...

This app is hosted on the root domain. So if a user hits `domain.com` it redirects to `domain.com/home` which is protected by the auth middleware. If the user is not logged in, they get redirected to the `/login` endpoint which hits the following method:

`public function login()
    {
        $authorize_params = [
            'scope' => 'openid profile email',
            // Use the key below to get an access token for your API.
            // 'audience' => config('laravel-auth0.api_identifier'),
        ];

        return \App::make('auth0')->login(null, null, $authorize_params);
    }`

At this point the app redirects them to the Auth0 hosted login page. 

The issue  I've identified is that if a user is already logged in and hits the `/login` end point it throws a `CoreException`. My solution is the following:

`public function login()
    {
        if(Auth::check())
        {
            return redirect()->route('home');
        }

        $authorize_params = [
            'scope' => 'openid profile email',
            // Use the key below to get an access token for your API.
            // 'audience' => config('laravel-auth0.api_identifier'),
        ];

        return \App::make('auth0')->login(null, null, $authorize_params);
    }`

Essentially I'm just doing a quick check and if the user is already authenticated then I redirect them back to the homepage. If not, it runs the Auth0 login flow.

Can you see any issues with this solution? Is this an issue unique to me or can it be recreated?

If this is acceptable I'll push to production and see if the CoreExceptions go away. If so, my thoughts are that some users might have their bookmark set to `domain.com/login` so if they try to use the bookmark and there's still a session active then it causes the CoreException.

**Comment by bkilshaw** [comment]
I'm assuming this is alright since it appears to be included in the latest version of the [Laravel Quickstart](https://auth0.com/docs/quickstart/webapp/laravel) guide!

**Comment by evansims** [comment]
Hey @bkilshaw 👋 Bingo! That would definitely cause an invalid state, since hitting the login() method issues new state and nonce cookies in transient storage. Give it a go and see if it solves things for you, I suspect it will.

**Comment by bkilshaw** [comment]
Hey @evansims 

So I'm still running into invalid state errors that I can't track down for the life of me.

Here's a gist of the exception trace: [https://gist.github.com/bkilshaw/81ba3fa73584d1cf1b11861507015e02](https://gist.github.com/bkilshaw/81ba3fa73584d1cf1b11861507015e02)

Even after fixing the login() issue I've still had 84 instances of this in the last 3 days.

I had one user that it was happening to consistently on an iPad (it happens on multiple different devices though) and even killing the browser and opening it up again didn't fix it, but it's back to working now.

**Comment by bkilshaw** [comment]
Is there anything else I can send over to help troubleshoot @evansims?

**Comment by bkilshaw** [comment]
@evansims Just checking in again here because I'm still getting the CoreExceptions "Invalid State" and I'm not too sure how to proceed.

**Comment by evansims** [comment]
Hey @bkilshaw 👋 Have you had any luck reproducing this issue on your own devices? This is a really difficult one to troubleshoot without either of us being able to reproduce.

The only instance where an invalid state will occur is if the device's stored state cookie is missing or doesn't match up with what's found when the user is returned to your callback. As you mentioned, hitting login() twice will invalid a state as it overwrites the cookie, which is a common one. That aside, there must be some sort of conflict occurring between what is being set on the end-users device from your application's login route (when login() with the SDK is invoked) and what's being found on the device when the end-user returns to your callback, and getUser()/exchange() is fired off.

Are you using getUser() elsewhere in your code that might be firing off on the callback route? getUser() invokes the code exchange, so it should only be getting fired once. When that fires, it deletes the temporary state cookie, so if it's getting fired off twice you might be invaliding the cookie.

**Comment by bkilshaw** [comment]
Still haven't been able to reproduce it! I've searched through all my code and I'm only calling getUser() once in my callback, and exchange() nowhere (the library handles it).

I'm a bit lost at how to continue. I'm contemplating moving to implement the php-sdk directly to see if that helps. I honestly don't really understand the inner workings of the auth0-laravel package so it makes it more difficult for me to figure it out.

I posted on the community looking to get a review of my implementation of the php-sdk but haven't heard back; hopefully someone can chime in and let me know if anything is missing or there's any security issues. 

[https://community.auth0.com/t/implementation-review/66161/2](https://community.auth0.com/t/implementation-review/66161/2)
